### PR TITLE
TB-4075: moving automation

### PR DESCRIPTION
**Pull request by vadpopov**
### What 🕵️ 🔍

- moving automated tests framework to main repo
- added lane for automation execution to fastfile

### CODE REVIEW

File: application/automation/src/main/java/com/xayn/screens/HomeScreen.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+import io.appium.java_client.pagefactory.iOSXCUITFindBy;
+import io.qameta.allure.Step;
+import lombok.extern.log4j.Log4j2;
+import org.openqa.selenium.WebDriverException;
+
+import java.util.Arrays;
+import java.util.List;
+
+@Log4j2
+public class HomeScreen extends BaseScreen {
```
Comment (joecks) [inline application/automation/src/main/java/com/xayn/screens/HomeScreen.java:16]: just for the future, this file can be generated from the existing dart file `Keys` now that our projects are living in the same space

File: application/automation/src/main/java/com/xayn/screens/ReaderModeScreen.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+import io.appium.java_client.MobileElement;
+import io.appium.java_client.pagefactory.AndroidFindBy;
+import io.appium.java_client.pagefactory.iOSXCUITFindBy;
+import io.qameta.allure.Step;
+
+import java.util.Arrays;
+
+public class ReaderModeScreen extends BaseScreen {
+
+    @AndroidFindBy(accessibility = "nav_bar_item_bookmark = true")
```
Comment (joecks) [inline application/automation/src/main/java/com/xayn/screens/ReaderModeScreen.java:14]: Ahh you are defining those accessibility id for every screen... hmm ok, isn't that a lot of repetition? Can't we use it in the BaseScreen and declare the values as protected?

File: application/automation/src/main/resources/config.properties [hunk 3]
```diff
@@ -0,0 +1,16 @@
+# All 'remote.*' params will be hardcoded for remote execution
+# Other params are dedicated for local execution on personal devices
+appium.url=http://localhost:4723/wd/hub
+screenshot.directory=target/screenshots
+wda.bundle.id=com.xayn.discovery.internal.wda
+retry.count=3
+tms.regression=false
+xcode.id=586TQ875ST
+remote.driver.exec=
```
Comment (joecks) [inline application/automation/src/main/resources/config.properties:9]: ❗ should be deleted and added to .gitignore, as discussed on the phone

File: application/automation/src/test/java/android/AndroidBookmarkingTest.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+                .clickOnHomeButton()
+                .bookmark()
+                .clickOnPersonalArea();
+        yourSpaceScreen.clickOnCollection(0);
+        onboarding.gotItButtonClick();
+        Assert.assertEquals(yourSpaceScreen.getAmountOfBookmarks(), 1, "Amount of bookmarks is different from expected");
+    }
+
+    @Test(description = "Adding a bookmark to Read Later collection from reading mode")
+    @TMS(id = 42)
```
Comment (joecks) [inline application/automation/src/test/java/android/AndroidBookmarkingTest.java:34]: How do you ensure that you havn't used an ID alreeady beforfe?

File: application/automation/src/test/java/android/AndroidCollectionsTest.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        String name =  RandomStringUtils.random(20, true, true);
+        component
+                .typeCollectionName(name)
+                .clickCreateButton();
+        yourSpaceScreen.clickOnCollection(1);
+        Assert.assertTrue(yourSpaceScreen.isNameOfTheCollectionDisplayed(name), "Name of the collection is different");
+    }
+
+    @Test(description = "Giving new collection the name that exceeds character limit of 20")
+    @TMS()
```
Comment (joecks) [inline application/automation/src/test/java/android/AndroidCollectionsTest.java:32]: 💅  no id?

File: application/automation/src/test/java/base/TestBase.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    public void beforeSuite(ITestContext context) {
+        ServerHandler.startLocalServer();
+        if (!ServerHandler.isServerRunning()) {
+            throw new RuntimeException("An appium server isn't running");
+        }
+        new File(Configuration.SCREENSHOT_DIRECTORY).mkdir();
+        // if (Configuration.IS_REGRESSION) TestRailHandler.createTestRun();
+//        for (ITestNGMethod method : context.getAllTestMethods()) {
+//            method.setRetryAnalyzerClass(new RetryAnalyzerImp());
+//        }
```
Comment (joecks) [inline application/automation/src/test/java/base/TestBase.java:30]: ❗  I guess this needs to be commented back?

File: application/automation/src/test/java/ios/IOSBookmarkingTest.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+import com.xayn.annotations.TMS;
+import com.xayn.screens.HomeScreen;
+import com.xayn.screens.YourSpaceScreen;
+import com.xayn.screens.components.OnboardingComponent;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+public class IOSBookmarkingTest extends IOSTestBase {
+    @Test(description = "Adding a bookmark to Read Later collection")
+    @TMS(id = 41)
```
Comment (joecks) [inline application/automation/src/test/java/ios/IOSBookmarkingTest.java:13]: 💅  this ID was already used before...

File: application/fastlane/Fastfile [hunk 8]
```diff
@@ -146,6 +174,13 @@ end
 #### Public lanes, can also be used via tab completion: `fastlane enable_auto_complete`
 ####
 
+desc "Run automation"
+lane :run_automation do |options|
+  set_up_automation_flag(options)
+  suite = options[Options::TEST_SUITE]
+  shell("cd automation && mvn clean test -DsuiteXmlFile=testng/" + suite + ".xml")
+end
```
Comment (joecks) [inline application/fastlane/Fastfile:182]: ❗  fastlane run_automation platform:ios test_suite:sanity

should:
- ensure ios simulator is setup and running
- build for ios simulatror
- run only for iOS

same for Android

File: application/fastlane/Fastfile [hunk 9]
```diff
@@ -146,6 +174,13 @@ end
 #### Public lanes, can also be used via tab completion: `fastlane enable_auto_complete`
 ####
 
+desc "Run automation"
+lane :run_automation do |options|
+  set_up_automation_flag(options)
+  suite = options[Options::TEST_SUITE]
```
Comment (joecks) [inline application/fastlane/Fastfile:180]: please add a default value in the `before_all` method

File: application/automation/allure-results/categories.json [hunk 10]
```diff
@@ -0,0 +1,10 @@
+[
+  {
+    "name": "Ignored tests",
+    "matchedStatuses": ["skipped"]
+  },
+  {
+    "name": "Product defects",
+    "matchedStatuses": ["failed"]
+  }
+]
```
[the remark this answers is not in the document]
Comment (vadpopov) [reply application/automation/allure-results/categories.json:10]: @joecks this one helps to filter results in report by assigning certain category to a test, we actually need it

File: application/automation/src/main/java/com/xayn/configuration/Configuration.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        XCODE_ORG_ID = props.getProperty("xcode.id");
+        IS_REGRESSION = Boolean.parseBoolean(props.getProperty("tms.regression"));
+        ANDROID_VERSION = getEnvProperty("android.version");
+        IOS_VERSION = getEnvProperty("ios.version");
+        IOS_DEVICE = getEnvProperty("ios.device");
+        //todo enable for automatic regression results
+//        TESTRAIL_ENDPOINT = System.getenv("TESTRAIL_ENDPOINT");
+//        TESTRAIL_USER = System.getenv("TESTRAIL_USER");
+//        TESTRAIL_PASSWORD = System.getenv("TESTRAIL_PASSWORD");
+        RETRY_COUNT = Integer.parseInt(props.getProperty("retry.count"));
```
[the remark this answers is not in the document]
Comment (vadpopov) [reply application/automation/src/main/java/com/xayn/configuration/Configuration.java:60]: @joecks we need those in order to add results to report automatically, but probably i can move all to somewhere else

File: application/automation/src/test/java/android/AndroidCollectionsTest.java (hunk 5 above)
Comment (vadpopov) [reply application/automation/src/test/java/android/AndroidCollectionsTest.java]: @joecks don't have a test case for it just yet, gonna add id later

File: application/automation/src/test/java/android/AndroidBookmarkingTest.java (hunk 4 above)
Comment (vadpopov) [reply application/automation/src/test/java/android/AndroidBookmarkingTest.java]: @joecks good point, probably can add some kind of a check for it

File: application/fastlane/lib/automation_utils.rb [hunk 12]
```diff
@@ -0,0 +1,14 @@
+def checkIfComponentsInstalled(params)
+    hash = {}
+    params.each do |i|
+        value = shell("which #{i} || echo ''")
+        puts value
```
Comment (joecks) [inline application/fastlane/lib/automation_utils.rb:5]: Better to use `Ui.success`

File: application/fastlane/lib/automation_utils.rb [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+        hash.store(i, value)
+        if value.empty?
+            UI.error "#{i} not found, did you forget to install it?"
```
Comment (joecks) [inline application/fastlane/lib/automation_utils.rb:8]: `UI.user_error!`

File: application/fastlane/Fastfile [hunk 14]
```diff
@@ -76,6 +79,64 @@ before_all do |lane, options|
 end
 
 # Private lanes, that we only call internally, they are nice because they have an own description
+
+desc "Checks properties required for automated execution locally"
+lane :check_automation_config do |options|
+  hash = checkIfComponentsInstalled(["appium","node","adb"])
```
Comment (joecks) [inline application/fastlane/Fastfile:85]: check also `mvn`

File: application/fastlane/Fastfile [hunk 15]
```diff
@@ -76,6 +79,64 @@ before_all do |lane, options|
...
+        deviceId = key["id"]
+        deviceVersion = /\d+/.match(key["sdk"])
+      end
+    end
+    if (deviceId.to_s.empty? || deviceVersion.to_s.empty?)
+      raise "Android device not found, have you enabled debug mode?"
+    end
+    #building the app
+    android_prepare_key_properties(options)
+    secrets_unpack_android(options)
```
Comment (joecks) [inline application/fastlane/Fastfile:103]: `secrets_unpack_android` probably not needed

File: application/automation/testng/debug.xml [hunk 16]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
+<suite name="Debug">
+    <listeners>
+        <listener class-name="com.xayn.listeners.ITestListenerImp"/>
+        <listener class-name="com.xayn.listeners.AnnotationTransformerImp"/>
+    </listeners>
+    <test name="Debug">
+        <classes>
+            <class name="android.AndroidNewsfeedTest"/>
```
Comment (joecks) [inline application/automation/testng/debug.xml:10]: this needs to be changed

File: application/fastlane/Fastfile [hunk 17]
```diff
@@ -76,6 +79,64 @@ before_all do |lane, options|
...
+      if key["targetPlatform"] == "ios"
+        deviceName = key["name"]
+        deviceVersion = /\d+-\d/.match(key["sdk"]).to_s.sub! '-', '.'
+        deviceId = key["id"]
+      end
+    end
+    if (deviceId.to_s.empty? || deviceVersion.to_s.empty?)
+      raise "IOS device not found"
+    end
+    shell("flutter build ios --simulator")
```
Comment (joecks) [inline application/fastlane/Fastfile:119]: ❗  please use `build_with_flutter(options)`
to make clear that you want to have a simulator build you could pass an internal param  (InternalOptions) 
to the options (`options  = options[InternalOptions::build_for_simulator] = true`)
and then read this out in the  build_with_flutter method to add the missing `--simulator` param

File: application/fastlane/Fastfile [hunk 18]
```diff
@@ -76,6 +79,64 @@ before_all do |lane, options|
...
+    app = File.expand_path("../build/ios/iphonesimulator/Runner.app");
+    raise ".app file isn't found" unless File.exist?app
```
Comment (joecks) [inline application/fastlane/Fastfile:121]: please use `Ui.user_error!` also `File.exist?app` seems to be wrong

File: application/fastlane/Fastfile [hunk 19]
```diff
@@ -76,6 +79,64 @@ before_all do |lane, options|
...
+  end
+  setProperties("../automation/src/main/resources/config.properties",
+                      {
+    "driver.exec" => hash["node"],
+    "appium.js" => hash["appium"],
+    "appium.url" => "http://localhost:4723/wd/hub",
+    "screenshot.directory" => "target/screenshots",
+    "wda.bundle.id" => "com.xayn.discovery.internal.wda",
+    "retry.count" => "3",
+    "testrail.app" => "Discovery App",
```
Comment (joecks) [inline application/fastlane/Fastfile:131]: 💅  if you want you could also pass the real name that is encoded in the config.rb hash

```
Flavors::BETA => {
      Platforms::IOS => {
        Keys::ID => "com.xayn.discovery",
        Keys::NAME => "Xayn News",
```

File: application/fastlane/Fastfile [hunk 20]
```diff
@@ -147,6 +208,14 @@ end
 #### Public lanes, can also be used via tab completion: `fastlane enable_auto_complete`
 ####
 
+# :test_suite and :test_env are required params for this lane
+desc "Run automation"
+lane :run_automation do |options|
+  check_automation_config(options)
+  suite = options[Options::TEST_SUITE]
+  shell("cd automation && mvn clean test -DsuiteXmlFile=testng/" + suite + ".xml")
```
Comment (joecks) [inline application/fastlane/Fastfile:216]:
```
Dir.chdir("../automation/") do
    sh(args)
end
```

File: application/fastlane/lib/config.rb [hunk 21]
```diff
@@ -209,4 +219,4 @@ module InternalOptions
   APPCENTER_DESTINATION_TYPE_STORE = "store"
   APPCENTER_DEFAULT_TARGET = "Collaborators"
   APPCENTER_DEFAULT_STORE = "Beta"
-end
+end
```
Comment (joecks) [inline application/fastlane/lib/config.rb:222]: 💅  please keep the line break at the end

File: application/fastlane/Fastfile (hunk 19 above)
Comment (vadpopov) [reply application/fastlane/Fastfile:131]: @joecks good point, will add later

File: application/fastlane/Fastfile [hunk 22]
```diff
@@ -63,6 +66,7 @@ before_all do |lane, options|
   setIfNotSet.call(Options::VERSION_NAME, version_name)
   setIfNotSet.call(Options::STORE, flavor == Flavors::BETA)
   options[_Options::ENV] = Config::ENV_DEFAULTS.to_h
+  options[_Options::BUILD_FOR_SIMULATOR] = platform == Platforms::IOS
```
Comment (joecks) [inline application/fastlane/Fastfile:69]: ❗  this is not true, most of the times we don't build for the simulator, please do not set this, unless we use it in the `run_automation` lane

File: application/fastlane/Fastfile [hunk 23]
```diff
@@ -76,6 +80,63 @@ before_all do |lane, options|
...
+      if key["targetPlatform"] == "ios"
+        deviceName = key["name"]
+        deviceVersion = /\d+-\d/.match(key["sdk"]).to_s.sub! '-', '.'
+        deviceId = key["id"]
+      end
+    end
+    if (deviceId.to_s.empty? || deviceVersion.to_s.empty?)
+      UI.user_error! "IOS device not found"
+    end
+    build_with_flutter(options)
```
Comment (joecks) [inline application/fastlane/Fastfile:119]: here you can set 
`options[_Options::BUILD_FOR_SIMULATOR] = true`

File: application/fastlane/lib/automation_utils.rb [hunk 24]
```diff
@@ -0,0 +1,14 @@
+def checkIfComponentsInstalled(params)
```
Comment (joecks) [inline application/fastlane/lib/automation_utils.rb:1]: I think this is a more general util, so maybe rename this to utils.rb

File: application/fastlane/lib/config.rb [hunk 25]
```diff
@@ -86,6 +92,9 @@ def self.doc_WATCH() "true, false" end
     def self.doc_TEST() "true, false" end
     TEST = :test
 
+    # def self.doc_BUILD_FOR_SIMULATOR() "true, false" end
+    #   BUILD_FOR_SIMULATOR = :build_for_simulator
+
```
Comment (joecks) [inline application/fastlane/lib/config.rb:97]: can be removed?

File: application/fastlane/lib/config.rb (hunk 21 above)
Comment (joecks) [reply application/fastlane/lib/config.rb:226]: still please keep the line break :P

File: application/fastlane/lib/config.rb (hunk 25 above)
Comment (vadpopov) [reply application/fastlane/lib/config.rb:97]: @joecks sure, sorry, missed that

**Review by joecks — APPROVED** [review]

**Pull request merged**